README.md
---
title: git-deb-pkg
section: 1
header: User manual
footer: git-deb-pkg
author:
  - Sébastien Gross
date: 2016-02-09 14:37:24
adjusting: b
hyphenate: yes
---

# NAME

git-deb-pkg - lazy way to build Debian package using git

# SYNOPSIS

git deb-pkg \[ *options* \] \[ commands \]

# DESCRIPTION

*git deb-pkg* offers a simple and easy way to manage Debian package
generation using git (and other source code management system).

Both upstream and Debian code are kept in separated branches.

Features:

  - Update upstream from any SCM which git is capable of.
  - Checkout any file from any branch.
  - Cherry pick any commit.
  - Few files to edit.
  - Keep **./debian** in a specific git branch.

# OPTIONS

## General options

-t DIR|--templates-dir DIR
: Directory where the *debhelper* files reside (default:
  **/usr/share/git-deb-pkg/templates**).

-h|--help
: Shows help screen.

## git options

-u BRANCH|--upstream-branch BRANCH
: Upstream branch (default: **upstream**).

-d BRANCH|--debian-branch BRANCH
: Debian branch (default: **debian**).

-U|--no-upstream-update
: Do not update upstream branch.

-C|--no-changelog-update
: Do not update debian changelog.

# COMMANDS

init-debian
: Init debian directory.

set-defaults
: Sets default options to git configuration.

commit-debian
: Commit the debian directory.

build
: Build debian package.

source
: Build debian source package only.

show-version
: Show all versions for the package.

update-upstream
: Update upstream branch.

prepare
: Prepare building environment.

prepare-orig
: Prepare upstream orig archive.

# CONFIGURATION OPTIONS

*git-deb-pkg* stores variables in git repository options.

## Variables

deb-pkg.scm
: Specify the source code management system to use. Default: **git**.

deb-pkg.upstream
: Specify the branch to use as upstream code. Default: **upstream**.

deb-pkg.debian
: Specify the branch to use for debian package definition. Default:
  **debian**.

deb-pkg.revision-scheme
: Regexp to extract version number from *git describe*. The git patchlevel
  *-PL-gTREEISH* is cut off before the regexp is applied. Default:
  **s/^v-\(.\*\)$/\1/;s/[-]/./g**. See UNDERSTANDING REVISION SCHEME.

deb-pkg.cherrypick
: Special commits to cherry pick. This option can occur many times. Order
  matters. Default: none.

deb-pkg.checkout
: Extra branch, directories or files to checkout before building the
  package.  This option can occur many times. Order matters. The syntax of
  each item is: *branch[:directory or file]*. Default: none.


# HOW TO USE

## Clone remote directory

First of all the upstream repository should be cloned the repository from a
remote source. Depending on the remote used SCM the commands may differ.

### git

    mkdir git-project
    cd git-project
    git remote add upstream git://remote.host/path/to/repository
    git fetch --all
    git checkout -b upstream upstream/master
    git checkout master

### svn

    mkdir git-project
    cd git-project
    git svn -i upstream clone clone http://remote.host/path/to/repository .

### cvs

    git cvsimport -d:pserver:anonymous@remote.host/path/to/repository \
        -r master -o master -a -C . -v module

## Create the **./debian** directory

This is where the magic starts. *git deb-pkg* helps a lot for that part:

	git deb-pkg set-defaults
    git deb-pkg init-debian

Package could be built using *debuild* command:

    debuild --no-tgz-check -I.git -i'\.git/'

To cancel and return to previous branch (assuming it was **master**) just run:

    debuild clean \
    	&& git reset --hard \
    	&& git checkout master \
    	&& git branch -D debian-dev

Then a few files in the *./debian* directory have to be updated, and commited:

    debuild clean
    git deb-pkg commit-debian

## build the package

To build the package from any branch, just type:

    git deb-pkg build

## build the source package

To build the source package from any branch, just type:

    git deb-pkg source

## publish package

This step is not mandatory unless you wish to publish your work:

    git remote add origin user@git.example.com:/path/to/repository.git
    git push -u origin master

## keep package up to date

### git

    git checkout master
    git fetch -a
    git merge upstream

### svn

    git checkout master
    git svn fetch
    git merge upstream

# UNDERSTANDING REVISION SCHEME
  
For example if the git tag is **v-1.2** and 5 commits before the **upstream**
branch, *git describe* would produce something like:

    v-1.2-5-gddf4ca0

Where:

- **v-1.2** is the current tag (understand upstram version)
- **5** is the number of commits after the current tag. This is the patch
   level (*PL*)
- **ddf5ca0** is the current commit ID prefixed with *g*. This is the git
   *TREEISH*.

The regexp defined in **deb-pkg.revision-scheme** variable is applied to the
current tag (**v-1.2**) and extracts the upstream version ('1.2').

The Debian package version would thus be: **1:1.2*git.5-1** which means:

- version *1.2* of the upstream
- git patch level *5*
- debian package revision *1*

See the Debian policy manual, controle files and their fields, Version
(http://www.debian.org/doc/debian-policy/ch-controlfields.html#s-f-Version)

# SEE ALSO

git(1), debhelper(7).

# HISTORY

2016/02/09
: Update template directory


2016/01/11
: create debian patch when modifing *autoconf.ac*

2012/01/10
: Add set-defaults option.

2011/12/09
: Fix DpkgVersion invocation.

2011/09/19
: Add source command.

2011/03/23
: Use git configuration file for options.

2011/03/22
: Change name from *debian-builder* to *git-deb-pkg* to git git tool suite.

2010/09/23
: Initial release.

# BUGS

No time to include bugs, command actions might seldom lead astray user's
assumption.


# COPYRIGHT

Copyright © 2010-2016 Sébastien Gross \<seb•ɑƬ•chezwam•ɖɵʈ•org\>.

Relesed under WTFPL (http://sam.zoy.org/wtfpl/COPYING).


debian/changelog
git-deb-pkg (1:1.7-1) unstable; urgency=low

  * Version 1.7

 -- Sébastien Gross <seb•ɑƬ•chezwam•ɖɵʈ•org>  Tue, 09 Feb 2016 15:09:26 +0200

git-deb-pkg (1:1.3+git.1-1) unstable; urgency=low

  * New automated build from: v1.3 - 0df0c2b

 -- Sébastien Gross <seb•ɑƬ•chezwam•ɖɵʈ•org>  Mon, 19 Sep 2011 11:57:53 +0200

git-deb-pkg (1:1.3-1) unstable; urgency=low

  * New automated build from: v1.3 - 8b5ad1d

 -- Sébastien Gross <seb•ɑƬ•chezwam•ɖɵʈ•org>  Mon, 19 Sep 2011 11:54:48 +0200

git-deb-pkg (1:1.3-1) unstable; urgency=low

  * New automated build from: v1.3 - 8b5ad1d

 -- Sébastien Gross <seb•ɑƬ•chezwam•ɖɵʈ•org>  Mon, 19 Sep 2011 11:45:06 +0200

git-deb-pkg (1.0-1) unstable; urgency=low

  * Initial release.

 -- Sébastien Gross <seb•ɑƬ•chezwam•ɖɵʈ•org>  Thu, 23 Sep 2010 15:15:18 +0000



templates/package.init.lsb.d
#!/bin/sh
#
# Example init.d script with LSB support.
#
# Please read this init.d carefully and modify the sections to
# adjust it to the program you want to run.
#
# Copyright (c) 2007 Javier Fernandez-Sanguino <jfs@debian.org>
#
# This is free software; you may redistribute it and/or modify
# it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 2,
# or (at your option) any later version.
#
# This is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License with
# the Debian operating system, in /usr/share/common-licenses/GPL;  if
# not, write to the Free Software Foundation, Inc., 59 Temple Place,
# Suite 330, Boston, MA 02111-1307 USA
#
### BEGIN INIT INFO
# Provides:          #PACKAGE#
# Required-Start:    $network $local_fs
# Required-Stop:
# Should-Start:      $named
# Should-Stop:
# Default-Start:     2 3 4 5
# Default-Stop:      0 1 6
# Short-Description: <Enter a short description of the sortware>
# Description:       <Enter a long description of the software>
#                    <...>
#                    <...>
### END INIT INFO

PATH=/usr/local/sbin:/usr/local/bin:/sbin:/bin:/usr/sbin:/usr/bin

DAEMON=/usr/sbin/#PACKAGE# # Introduce the server's location here
NAME=#PACKAGE              # Introduce the short server's name here
DESC=#PACKAGE              # Introduce a short description here
LOGDIR=/var/log/#PACKAGE#  # Log directory to use

PIDFILE=/var/run/$NAME.pid

test -x $DAEMON || exit 0

. /lib/lsb/init-functions

# Default options, these can be overriden by the information
# at /etc/default/$NAME
DAEMON_OPTS=""          # Additional options given to the server

DIETIME=10              # Time to wait for the server to die, in seconds
                        # If this value is set too low you might not
                        # let some servers to die gracefully and
                        # 'restart' will not work

#STARTTIME=2             # Time to wait for the server to start, in seconds
                        # If this value is set each time the server is
                        # started (on start or restart) the script will
                        # stall to try to determine if it is running
                        # If it is not set and the server takes time
                        # to setup a pid file the log message might 
                        # be a false positive (says it did not start
                        # when it actually did)
                        
LOGFILE=$LOGDIR/$NAME.log  # Server logfile
#DAEMONUSER=#PACKAGE#   # Users to run the daemons as. If this value
                        # is set start-stop-daemon will chuid the server

# Include defaults if available
if [ -f /etc/default/$NAME ] ; then
	. /etc/default/$NAME
fi

# Use this if you want the user to explicitly set 'RUN' in
# /etc/default/
#if [ "x$RUN" != "xyes" ] ; then
#    log_failure_msg "$NAME disabled, please adjust the configuration to your needs "
#    log_failure_msg "and then set RUN to 'yes' in /etc/default/$NAME to enable it."
#    exit 1
#fi

# Check that the user exists (if we set a user)
# Does the user exist?
if [ -n "$DAEMONUSER" ] ; then
    if getent passwd | grep -q "^$DAEMONUSER:"; then
        # Obtain the uid and gid
        DAEMONUID=`getent passwd |grep "^$DAEMONUSER:" | awk -F : '{print $3}'`
        DAEMONGID=`getent passwd |grep "^$DAEMONUSER:" | awk -F : '{print $4}'`
    else
        log_failure_msg "The user $DAEMONUSER, required to run $NAME does not exist."
        exit 1
    fi
fi


set -e

running_pid() {
# Check if a given process pid's cmdline matches a given name
    pid=$1
    name=$2
    [ -z "$pid" ] && return 1
    [ ! -d /proc/$pid ] &&  return 1
    cmd=`cat /proc/$pid/cmdline | tr "\000" "\n"|head -n 1 |cut -d : -f 1`
    # Is this the expected server
    [ "$cmd" != "$name" ] &&  return 1
    return 0
}

running() {
# Check if the process is running looking at /proc
# (works for all users)

    # No pidfile, probably no daemon present
    [ ! -f "$PIDFILE" ] && return 1
    pid=`cat $PIDFILE`
    running_pid $pid $DAEMON || return 1
    return 0
}

start_server() {
# Start the process using the wrapper
        if [ -z "$DAEMONUSER" ] ; then
            start_daemon -p $PIDFILE $DAEMON -- $DAEMON_OPTS
            errcode=$?
        else
# if we are using a daemonuser then change the user id
            start-stop-daemon --start --quiet --pidfile $PIDFILE \
                        --chuid $DAEMONUSER \
                        --exec $DAEMON -- $DAEMON_OPTS
            errcode=$?
        fi
	return $errcode
}

stop_server() {
# Stop the process using the wrapper
        if [ -z "$DAEMONUSER" ] ; then
            killproc -p $PIDFILE $DAEMON
            errcode=$?
        else
# if we are using a daemonuser then look for process that match
            start-stop-daemon --stop --quiet --pidfile $PIDFILE \
                        --user $DAEMONUSER \
                        --exec $DAEMON
            errcode=$?
        fi

	return $errcode
}

reload_server() {
    [ ! -f "$PIDFILE" ] && return 1
    pid=pidofproc $PIDFILE # This is the daemon's pid
    # Send a SIGHUP
    kill -1 $pid
    return $?
}

force_stop() {
# Force the process to die killing it manually
	[ ! -e "$PIDFILE" ] && return
	if running ; then
		kill -15 $pid
	# Is it really dead?
		sleep "$DIETIME"s
		if running ; then
			kill -9 $pid
			sleep "$DIETIME"s
			if running ; then
				echo "Cannot kill $NAME (pid=$pid)!"
				exit 1
			fi
		fi
	fi
	rm -f $PIDFILE
}


case "$1" in
  start)
	log_daemon_msg "Starting $DESC " "$NAME"
        # Check if it's running first
        if running ;  then
            log_progress_msg "apparently already running"
            log_end_msg 0
            exit 0
        fi
        if start_server ; then
            # NOTE: Some servers might die some time after they start,
            # this code will detect this issue if STARTTIME is set
            # to a reasonable value
            [ -n "$STARTTIME" ] && sleep $STARTTIME # Wait some time 
            if  running ;  then
                # It's ok, the server started and is running
                log_end_msg 0
            else
                # It is not running after we did start
                log_end_msg 1
            fi
        else
            # Either we could not start it
            log_end_msg 1
        fi
	;;
  stop)
        log_daemon_msg "Stopping $DESC" "$NAME"
        if running ; then
            # Only stop the server if we see it running
			errcode=0
            stop_server || errcode=$?
            log_end_msg $errcode
        else
            # If it's not running don't do anything
            log_progress_msg "apparently not running"
            log_end_msg 0
            exit 0
        fi
        ;;
  force-stop)
        # First try to stop gracefully the program
        $0 stop
        if running; then
            # If it's still running try to kill it more forcefully
            log_daemon_msg "Stopping (force) $DESC" "$NAME"
			errcode=0
            force_stop || errcode=$?
            log_end_msg $errcode
        fi
	;;
  restart|force-reload)
        log_daemon_msg "Restarting $DESC" "$NAME"
		errcode=0
        stop_server || errcode=$?
        # Wait some sensible amount, some server need this
        [ -n "$DIETIME" ] && sleep $DIETIME
        start_server || errcode=$?
        [ -n "$STARTTIME" ] && sleep $STARTTIME
        running || errcode=$?
        log_end_msg $errcode
	;;
  status)

        log_daemon_msg "Checking status of $DESC" "$NAME"
        if running ;  then
            log_progress_msg "running"
            log_end_msg 0
        else
            log_progress_msg "apparently not running"
            log_end_msg 1
            exit 1
        fi
        ;;
  # Use this if the daemon cannot reload
  reload)
        log_warning_msg "Reloading $NAME daemon: not implemented, as the daemon"
        log_warning_msg "cannot re-read the config file (use restart)."
        ;;
  # And this if it cann
  #reload)
          #
          # If the daemon can reload its config files on the fly
          # for example by sending it SIGHUP, do it here.
          #
          # If the daemon responds to changes in its config file
          # directly anyway, make this a do-nothing entry.
          #
          # log_daemon_msg "Reloading $DESC configuration files" "$NAME"
          # if running ; then
          #    reload_server
          #    if ! running ;  then
          # Process died after we tried to reload
          #       log_progress_msg "died on reload"
          #       log_end_msg 1
          #       exit 1
          #    fi
          # else
          #    log_progress_msg "server is not running"
          #    log_end_msg 1
          #    exit 1
          # fi
                                                                                    #;;

  *)
	N=/etc/init.d/$NAME
	echo "Usage: $N {start|stop|force-stop|restart|force-reload|status}" >&2
	exit 1
	;;
esac

exit 0


templates/package.init.d
#! /bin/sh
#
# skeleton	example file to build /etc/init.d/ scripts.
#		This file should be used to construct scripts for /etc/init.d.
#
#		Written by Miquel van Smoorenburg <miquels@cistron.nl>.
#		Modified for Debian
#		by Ian Murdock <imurdock@gnu.ai.mit.edu>.
#               Further changes by Javier Fernandez-Sanguino <jfs@debian.org>
#
# Version:	@(#)skeleton  1.9  26-Feb-2001  miquels@cistron.nl
#

PATH=/usr/local/sbin:/usr/local/bin:/sbin:/bin:/usr/sbin:/usr/bin
DAEMON=/usr/sbin/#PACKAGE#
NAME=#PACKAGE#
DESC=#PACKAGE#

test -x $DAEMON || exit 0

LOGDIR=/var/log/#PACKAGE#
PIDFILE=/var/run/$NAME.pid
DODTIME=1                   # Time to wait for the server to die, in seconds
                            # If this value is set too low you might not
                            # let some servers to die gracefully and
                            # 'restart' will not work

# Include #PACKAGE# defaults if available
if [ -f /etc/default/#PACKAGE# ] ; then
	. /etc/default/#PACKAGE#
fi

set -e

running_pid()
{
    # Check if a given process pid's cmdline matches a given name
    pid=$1
    name=$2
    [ -z "$pid" ] && return 1
    [ ! -d /proc/$pid ] &&  return 1
    cmd=`cat /proc/$pid/cmdline | tr "\000" "\n"|head -n 1 |cut -d : -f 1`
    # Is this the expected child?
    [ "$cmd" != "$name" ] &&  return 1
    return 0
}

running()
{
# Check if the process is running looking at /proc
# (works for all users)

    # No pidfile, probably no daemon present
    [ ! -f "$PIDFILE" ] && return 1
    # Obtain the pid and check it against the binary name
    pid=`cat $PIDFILE`
    running_pid $pid $DAEMON || return 1
    return 0
}

force_stop() {
# Forcefully kill the process
    [ ! -f "$PIDFILE" ] && return
    if running ; then
        kill -15 $pid
        # Is it really dead?
        [ -n "$DODTIME" ] && sleep "$DODTIME"s
        if running ; then
            kill -9 $pid
            [ -n "$DODTIME" ] && sleep "$DODTIME"s
            if running ; then
                echo "Cannot kill $LABEL (pid=$pid)!"
                exit 1
            fi
        fi
    fi
    rm -f $PIDFILE
    return 0
}

case "$1" in
  start)
	echo -n "Starting $DESC: "
	start-stop-daemon --start --quiet --pidfile $PIDFILE \
		--exec $DAEMON -- $DAEMON_OPTS
        if running ; then
            echo "$NAME."
        else
            echo " ERROR."
        fi
	;;
  stop)
	echo -n "Stopping $DESC: "
	start-stop-daemon --stop --quiet --pidfile $PIDFILE \
		--exec $DAEMON
	echo "$NAME."
	;;
  force-stop)
	echo -n "Forcefully stopping $DESC: "
        force_stop
        if ! running ; then
            echo "$NAME."
        else
            echo " ERROR."
        fi
	;;
  #reload)
	#
	#	If the daemon can reload its config files on the fly
	#	for example by sending it SIGHUP, do it here.
	#
	#	If the daemon responds to changes in its config file
	#	directly anyway, make this a do-nothing entry.
	#
	# echo "Reloading $DESC configuration files."
	# start-stop-daemon --stop --signal 1 --quiet --pidfile \
	#	/var/run/$NAME.pid --exec $DAEMON
  #;;
  force-reload)
	#
	#	If the "reload" option is implemented, move the "force-reload"
	#	option to the "reload" entry above. If not, "force-reload" is
	#	just the same as "restart" except that it does nothing if the
	#   daemon isn't already running.
	# check wether $DAEMON is running. If so, restart
	start-stop-daemon --stop --test --quiet --pidfile \
		/var/run/$NAME.pid --exec $DAEMON \
	&& $0 restart \
	|| exit 0
	;;
  restart)
    echo -n "Restarting $DESC: "
	start-stop-daemon --stop --quiet --pidfile \
		/var/run/$NAME.pid --exec $DAEMON
	[ -n "$DODTIME" ] && sleep $DODTIME
	start-stop-daemon --start --quiet --pidfile \
		/var/run/$NAME.pid --exec $DAEMON -- $DAEMON_OPTS
	echo "$NAME."
	;;
  status)
    echo -n "$LABEL is "
    if running ;  then
        echo "running"
    else
        echo " not running."
        exit 1
    fi
    ;;
  *)
	N=/etc/init.d/$NAME
	# echo "Usage: $N {start|stop|restart|reload|force-reload}" >&2
	echo "Usage: $N {start|stop|restart|force-reload|status|force-stop}" >&2
	exit 1
	;;
esac

exit 0
